src/Adapter/MSTest.TestAdapter/Execution/TestMethodInfo.cs

```diff
@@ -467,65 +468,84 @@ private void RunTestCleanupMethod(object classInstance, TestResult result)
-                    Resource.UTA_CleanupMethodThrows,
-                    TestClassName,
-                    testCleanupMethod.Name,
-                    formattedExceptionMessage);
-            }
-            else
+        var cleanupError = new StringBuilder();
+        var cleanupStackTrace = new StringBuilder();
+        if (result.TestFailureException is TestFailedException testFailureException)
+        {
+            if (!StringEx.IsNullOrEmpty(testFailureException.Message))
             {
-                cleanupError.AppendFormat(
-                    CultureInfo.CurrentCulture,
-                    Resource.UTA_CleanupMethodThrowsGeneralError,
-                    TestClassName,
-                    formattedExceptionMessage);
+                cleanupError.Append(testFailureException.Message);
+                cleanupError.AppendLine();
             }
 
-            StackTraceInformation? cleanupStackTraceInfo = null;
-            var realExceptionStackTraceInfo = realException.TryGetStackTraceInformation();
-            if (realExceptionStackTraceInfo != null)
+            if (!StringEx.IsNullOrEmpty(testFailureException.StackTraceInformation?.ErrorStackTrace))
             {
-                cleanupStackTrace.Append(realExceptionStackTraceInfo.ErrorStackTrace);
-                cleanupStackTraceInfo ??= realExceptionStackTraceInfo;
+                cleanupStackTrace.Append(testFailureException.StackTraceInformation.ErrorStackTrace);
+                cleanupStackTrace.Append(Environment.NewLine);
+                cleanupStackTrace.Append(Resource.UTA_CleanupStackTrace);
+                cleanupStackTrace.Append(Environment.NewLine);
             }
+        }
+
+        Exception realException = testCleanupException.GetRealException();
+        string formattedExceptionMessage = realException.GetFormattedExceptionMessage();
+
+        if (testCleanupMethod != null)
+        {
+            cleanupError.AppendFormat(
+                CultureInfo.CurrentCulture,
+                Resource.UTA_CleanupMethodThrows,
+                TestClassName,
+                testCleanupMethod.Name,
+                formattedExceptionMessage);
+        }
+        else
+        {
+            cleanupError.AppendFormat(
+                CultureInfo.CurrentCulture,
+                Resource.UTA_CleanupMethodThrowsGeneralError,
+                TestClassName,
+                formattedExceptionMessage);
+        }
 
-            var finalStackTraceInfo = cleanupStackTraceInfo != null
+        StackTraceInformation? cleanupStackTraceInfo = null;
+        var realExceptionStackTraceInfo = realException.TryGetStackTraceInformation();
+        if (realExceptionStackTraceInfo != null)
+        {
+            cleanupStackTrace.Append(realExceptionStackTraceInfo.ErrorStackTrace);
+            cleanupStackTraceInfo ??= realExceptionStackTraceInfo;
+        }
+
+        StackTraceInformation? finalStackTraceInfo = null;
```

Comment:
this can be now null, in past wasn't possible are we sure that's fine?

Reply:
we have tests with the stack empty and no exception happend